File: algorithms/152_maximum_product_subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        if (nums.empty()) return 0;
        int min_product=nums[0],max_product=nums[0], maximum=nums[0];
        for (int i=1;i<nums.size();i++) {
            int t1=min(min_product*nums[i], max_product*nums[i]);
            int t2=max(min_product*nums[i], max_product*nums[i]);
            min_product=min(t1,nums[i]);
            max_product=max(t2,nums[i]);
            maximum=maximum>max_product? maximum:max_product;
        }
        return maximum;
    }
};

File: algorithms/python3/152_maximum_product_subarray.py
from typing import List
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0
        record=[nums[0]]*2
        res=nums[0]

        for cur in nums[1:]:
            record=[min(record[0]*cur,cur,record[1]*cur),max(record[0]*cur,cur,record[1]*cur)]
            res=max(res,record[1])
        return res

a=Solution()

nums=[2,3,-2,4]
print(a.maxProduct(nums))